File: PaleoRose/Classes/Layer Control/XRLayer.m
//
//  XRLayer.m
//  XRose
//
//  Created by Tom Moore on Sat Jan 24 2004.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "XRLayer.h"
#import "XRGeometryController.h"
#import "XRLayerCore.h"
#import "XRLayerGrid.h"
#import "XRLayerData.h"
#import "XRLayerText.h"
#import "XRLayerLineArrow.h"
#import <PaleoRose-Swift.h>

@implementation XRLayer

-(id)init {
    self = [super init];
    if(self) {
        _graphicalObjects = [[NSMutableArray alloc] init];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChange:) name:XRGeometryDidChange object:geometryController];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangePercent:) name:XRGeometryDidChangeIsPercent object:geometryController];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangeSectors:) name:XRGeometryDidChangeSectors object:geometryController];
        [self setStrokeColor:[NSColor blackColor]];
        [self setFillColor:[NSColor blackColor]];
        [self setIsVisible:YES];
        [self setIsActive:NO];
        _canFill = YES;
        _canStroke = YES;
    }
    return self;
}

-(id)initWithGeometryController:(XRGeometryController *)aController
{
	if (!(self = [super init])) return nil;
	if(self)
	{
		geometryController = aController;
		_graphicalObjects = [[NSMutableArray alloc] init];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChange:) name:XRGeometryDidChange object:geometryController];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangePercent:) name:XRGeometryDidChangeIsPercent object:geometryController];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangeSectors:) name:XRGeometryDidChangeSectors object:geometryController];
		[self setStrokeColor:[NSColor blackColor]];
		[self setFillColor:[NSColor blackColor]];
		[self setIsVisible:YES];
		[self setIsActive:NO];
		_canFill = YES;
		_canStroke = YES;
		
			}
	return self;
}

-(id)initWithGeometryController:(XRGeometryController *)aController dictionary:(NSDictionary *)configure
{
	if (!(self = [super init])) return nil;
	if(self)
	{
		NSColor *aColor;
		NSString *tempString;
		geometryController = aController;
		_graphicalObjects = [[NSMutableArray alloc] init];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChange:) name:XRGeometryDidChange object:geometryController];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangePercent:) name:XRGeometryDidChangeIsPercent object:geometryController];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangeSectors:) name:XRGeometryDidChangeSectors object:geometryController];
		if((aColor = [configure objectForKey:@"Stroke_Color"]))
			[self setStrokeColor:aColor];
		if((aColor = [configure objectForKey:@"Fill_Color"]))
			[self setFillColor:aColor];
		
		if((tempString = [configure objectForKey:@"Visible"]))
		{
			if([tempString isEqualToString:@"YES"])
				[self setIsVisible:YES];
			else
				[self setIsVisible:NO];
		}
		if((tempString = [configure objectForKey:@"Active"]))
		{
			if([tempString isEqualToString:@"YES"])
				[self setIsActive:YES];
			else
				[self setIsActive:NO];
		}
		
		if((tempString = [configure objectForKey:@"BIDIR"]))
		{
			if([tempString isEqualToString:@"YES"])
				_isBiDir = YES;
			else
				_isBiDir = NO;
		}
		
		if((tempString = [configure objectForKey:@"Layer_Name"]))
			_layerName = tempString;
		
		if((tempString = [configure objectForKey:@"Line_Weight"]))
			_lineWeight = [tempString floatValue];
		
		if((tempString = [configure objectForKey:@"Max_Count"]))
			_maxCount = [tempString floatValue];
		
		if((tempString = [configure objectForKey:@"Max_Percent"]))
			_lineWeight = [tempString floatValue];
		
		_canFill = YES;
		_canStroke = YES;
		
		 
	}
	return self;
}

-(void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

-(BOOL)isVisible
{
	return _isVisible;
}

-(void)setIsVisible:(BOOL)visible
{
	if(visible==_isVisible)
		return;
	_isVisible = visible;
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];
}

-(BOOL)isActive
{
	return _isActive;
}

-(void)setIsActive:(BOOL)active
{
	_isActive = active;//possibly change if needed a UI reaction
}

-(NSColor *)strokeColor
{
	return _strokeColor;
}

-(void)setStrokeColor:(NSColor *)color
{
	NSEnumerator *anEnum = [_graphicalObjects objectEnumerator];
	Graphic *aGraphic;
	_strokeColor = color;
	[self resetColorImage];
	//possibly post notification.
	while(aGraphic = [anEnum nextObject])
	{
		[aGraphic setStrokeColor:_strokeColor];
	}
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];

}

-(NSColor *)fillColor
{
	return _fillColor;
}

-(void)setFillColor:(NSColor *)color
{
	NSEnumerator *anEnum = [_graphicalObjects objectEnumerator];
	Graphic *aGraphic;
	_fillColor = color;
	[self resetColorImage];
	//possibly post notification.
	while(aGraphic = [anEnum nextObject])
	{
        if([aGraphic respondsToSelector:@selector(setFillColor:)]) {
            [aGraphic setFillColor:_fillColor];
        }
	}
	//NSLog(@"set fill color %@",[[self class] description]);
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];
}

-(NSString *)layerName
{
	return _layerName;
}

-(void)setLayerName:(NSString *)layerName
{
	_layerName = layerName;
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerTableRequiresReload object:self];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];
}

-(BOOL)isBiDirectional
{
	return _isBiDir;
}

-(void)setBiDirectional:(BOOL)isBiDir
{

	_isBiDir = isBiDir;
	//requires updating the counts and percents
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerInspectorRequiresReload object:self];
}

-(void)resetColorImage
{

	NSRect imageRect = NSMakeRect(0.0,0.0,16.0,16.0);
	_anImage = nil;
	_anImage = [[NSImage alloc] initWithSize:imageRect.size];
	//if(_anImage)
	//	NSLog([_anImage description]);
	//NSLog(@"reset image for object %@",[self layerName]);
	//NSLog(@"resetColorImage 1");
	NSBezierPath *aPath = [NSBezierPath bezierPathWithRect:imageRect];
	//NSLog(@"resetColorImage 1.1");
	[aPath setLineWidth:4.0];
	//NSLog(@"resetColorImage 1.2");
	[_anImage lockFocus];
	//NSLog(@"resetColorImage 1.3");
	[_strokeColor set];
	//NSLog(@"resetColorImage 1.4");
	[aPath stroke];
	//NSLog(@"resetColorImage 2");
	imageRect = NSInsetRect(imageRect,2.0,2.0);
	//NSLog(@"resetColorImage %i",[_graphicalObjects count]);
	if([_graphicalObjects count] > 0)
	{
		if([[_graphicalObjects objectAtIndex:0] respondsToSelector:@selector(drawsFill)] && [[_graphicalObjects objectAtIndex:0] drawsFill])
		{
			//NSLog(@"draws fill");
			[_fillColor set];
			aPath = [NSBezierPath bezierPathWithRect:imageRect];
			[aPath fill];
		}
	}
	//NSLog(@"resetColorImage 3");
	[_anImage unlockFocus];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerTableRequiresReload object:self];

	
		

}
-(NSImage *)colorImage
{
	return _anImage;
}

-(void)generateGraphics
{
	
}

-(void)drawRect:(NSRect)rect
{
}

-(void)geometryDidChange:(NSNotification *)notification
{
	[self generateGraphics];

	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
}

-(void)geometryDidChangePercent:(NSNotification *)notification
{
	[self generateGraphics];
	
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
}

-(void)geometryDidChangeSectors:(NSNotification *)notification
{
	[self generateGraphics];
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
}

-(int)maxCount
{
	return _maxCount;
}

-(float)maxPercent
{
	return _maxPercent;
}

-(void)setLineWeight:(float)lineWeight
{
	_lineWeight = lineWeight;
	[[NSNotificationCenter defaultCenter] postNotificationName:XRLayerRequiresRedraw object:self];
}

-(float)lineWeight
{
	return _lineWeight;
}

-(NSDictionary *)layerSettings
{
	NSMutableDictionary *aDict = [[NSMutableDictionary alloc] init];
	//NSLog(@"standard layer");
	if([self isKindOfClass:[XRLayerCore class]])
		[aDict setObject:@"Core_Layer" forKey:XRLayerXMLType];
	else if([self isKindOfClass:[XRLayerData class]])
		[aDict setObject:@"Data_Layer" forKey:XRLayerXMLType];
	else if([self isKindOfClass:[XRLayerGrid class]])
		[aDict setObject:@"Grid_Layer" forKey:XRLayerXMLType];
	else
		[aDict setObject:@"Layer" forKey:XRLayerXMLType];
	
	if(_isVisible)
		[aDict setObject:@"YES" forKey:@"Visible"];
	else
		[aDict setObject:@"NO" forKey:@"Visible"];
	
	if(_isActive)
		[aDict setObject:@"YES" forKey:@"Active"];
	else
		[aDict setObject:@"NO" forKey:@"Active"];
	
	[aDict setObject:_layerName forKey:@"Layer_Name"];
	
	if(_isBiDir)
		[aDict setObject:@"YES" forKey:@"BIDIR"];
	else
		[aDict setObject:@"NO" forKey:@"BIDIR"];

	[aDict setObject:_strokeColor forKey:@"Stroke_Color"];

	[aDict setObject:_fillColor forKey:@"Fill_Color"];
	//NSLog(@"line weight %f",_lineWeight);
	[aDict setObject:[NSString stringWithFormat:@"%f",_lineWeight] forKey:@"Line_Weight"];
	[aDict setObject:[NSString stringWithFormat:@"%i",_maxCount] forKey:@"Max_Count"];
	[aDict setObject:[NSString stringWithFormat:@"%f",_maxPercent] forKey:@"Max_Percent"];
	//NSLog(@"end standard layer");
	return [NSDictionary dictionaryWithDictionary:aDict];
}

-(XRDataSet *)dataSet
{
	return nil;
}

-(void)setDataSet:(XRDataSet *)aSet
{
	return;
}

-(BOOL)handleMouseEvent:(NSEvent *)anEvent
{
	//NSLog(@"layer handle mouse down");
	return NO;
}

-(BOOL)hitDetection:(NSPoint)testPoint
{
	return NO;
}

-(void)setGeometryController:(XRGeometryController *)controller
{
	if (geometryController == controller) {
		return;
	}

	// Remove old observers if we had a previous controller
	if (geometryController) {
		[[NSNotificationCenter defaultCenter] removeObserver:self name:XRGeometryDidChange object:geometryController];
		[[NSNotificationCenter defaultCenter] removeObserver:self name:XRGeometryDidChangeIsPercent object:geometryController];
		[[NSNotificationCenter defaultCenter] removeObserver:self name:XRGeometryDidChangeSectors object:geometryController];
	}

	geometryController = controller;

	// Add new observers
	if (geometryController) {
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChange:) name:XRGeometryDidChange object:geometryController];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangePercent:) name:XRGeometryDidChangeIsPercent object:geometryController];
		[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(geometryDidChangeSectors:) name:XRGeometryDidChangeSectors object:geometryController];

		// Generate graphics now that we have a geometry controller
		// BUT: XRLayerData and XRLayerLineArrow need their dataset set first,
		// so they should NOT generate graphics here. Their setDataSet method will do it.
		if (![self isKindOfClass:[XRLayerData class]] && ![self isKindOfClass:[XRLayerLineArrow class]]) {
			[self generateGraphics];
		}
	}
}

-(XRGeometryController *)geometryController
{
	return geometryController;
}

-(NSString *)type
{
    return @"XRLayer";
}

@end

File: PaleoRose/Classes/Layer Control/XRLayer.swift
// XRLayer.swift
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Cocoa
import SQLite3

// MARK: - Constants

let XRLayerRequiresRedraw = "XRLayerRequiresRedraw"
let XRLayerTableRequiresReload = "XRLayerTableRequiresReload"
let XRLayerInspectorRequiresReload = "XRLayerInspectorRequiresReload"

let XRLayerXMLType = "Layer_Type"
let XRLayerGraphicObjectArray = "Graphics"

@objc class XRLayer: NSObject {
    // MARK: - Properties

    private var graphicalObjects: [XRGraphic] = []
    private var isVisible: Bool = true {
        didSet {
            if oldValue != isVisible {
                NotificationCenter.default.post(name: NSNotification.Name(XRLayerRequiresRedraw), object: self)
                NotificationCenter.default.post(name: NSNotification.Name(XRLayerInspectorRequiresReload), object: self)
            }
        }
    }

    private var isActive: Bool = false
    private var strokeColor: NSColor = .black {
        didSet {
            resetColorImage()
            graphicalObjects.forEach { $0.strokeColor = strokeColor }
            NotificationCenter.default.post(name: NSNotification.Name(XRLayerRequiresRedraw), object: self)
            NotificationCenter.default.post(name: NSNotification.Name(XRLayerInspectorRequiresReload), object: self)
        }
    }

    private var fillColor: NSColor = .black {
        didSet {
            resetColorImage()
            graphicalObjects.forEach { $0.fillColor = fillColor }
            NotificationCenter.default.post(name: NSNotification.Name(XRLayerRequiresRedraw), object: self)
            NotificationCenter.default.post(name: NSNotification.Name(XRLayerInspectorRequiresReload), object: self)
        }
    }

    private var layerName: String = ""
    private var isBiDir: Bool = false
    private var lineWeight: Float = 1.0
    private var maxCount: Int = 0 // Should be set when sector size is set
    private var maxPercent: Float = 0.0
    private var anImage: NSImage?
    private var canFill: Bool = true
    private var canStroke: Bool = true
    private weak var geometryController: XRGeometryController?
    private weak var dataSet: XRDataSet?

    // MARK: - Class Methods

    @objc class func classTag() -> String {
        "Layer"
    }

    // MARK: - Initialization

    @objc override init() {
        super.init()
        setupNotifications()
        setupDefaults()
    }

    @objc init(geometryController: XRGeometryController) {
        super.init()
        self.geometryController = geometryController
        setupNotifications()
        setupDefaults()
    }

    @objc init(geometryController: XRGeometryController, dictionary: [String: Any]) {
        super.init()
        self.geometryController = geometryController
        setupNotifications()
        configureSelf(with: dictionary)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    // MARK: - Setup Methods

    private func setupNotifications() {
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(geometryDidChange(_:)),
                                               name: NSNotification.Name("XRGeometryDidChange"),
                                               object: geometryController)
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(geometryDidChangePercent(_:)),
                                               name: NSNotification.Name("XRGeometryDidChangeIsPercent"),
                                               object: geometryController)
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(geometryDidChangeSectors(_:)),
                                               name: NSNotification.Name("XRGeometryDidChangeSectors"),
                                               object: geometryController)
    }

    private func setupDefaults() {
        strokeColor = .black
        fillColor = .black
        isVisible = true
        isActive = false
        canFill = true
        canStroke = true
    }

    private func configureSelf(with dictionary: [String: Any]) {
        if let color = dictionary["Stroke_Color"] as? NSColor {
            strokeColor = color
        }
        if let color = dictionary["Fill_Color"] as? NSColor {
            fillColor = color
        }
        if let visibleStr = dictionary["Visible"] as? String {
            isVisible = (visibleStr == "YES")
        }
        if let activeStr = dictionary["Active"] as? String {
            isActive = (activeStr == "YES")
        }
        if let bidirStr = dictionary["BIDIR"] as? String {
            isBiDir = (bidirStr == "YES")
        }
        if let name = dictionary["Layer_Name"] as? String {
            layerName = name
        }
        if let weight = dictionary["Line_Weight"] as? String {
            lineWeight = Float(weight) ?? 1.0
        }
        if let count = dictionary["Max_Count"] as? String {
            maxCount = Int(count) ?? 0
        }
        if let percent = dictionary["Max_Percent"] as? String {
            maxPercent = Float(percent) ?? 0.0
        }
    }

    // MARK: - Public Methods

    @objc func getIsVisible() -> Bool {
        isVisible
    }

    @objc func setIsVisible(_ visible: Bool) {
        isVisible = visible
    }

    @objc func getIsActive() -> Bool {
        isActive
    }

    @objc func setIsActive(_ active: Bool) {
        isActive = active
    }

    @objc func getStrokeColor() -> NSColor {
        strokeColor
    }

    @objc func setStrokeColor(_ color: NSColor) {
        strokeColor = color
    }

    @objc func getFillColor() -> NSColor {
        fillColor
    }

    @objc func setFillColor(_ color: NSColor) {
        fillColor = color
    }

    @objc func getLayerName() -> String {
        layerName
    }

    @objc func setLayerName(_ name: String) {
        layerName = name
    }

    @objc func getIsBiDirectional() -> Bool {
        isBiDir
    }

    @objc func setBiDirectional(_ biDir: Bool) {
        isBiDir = biDir
    }

    @objc func resetColorImage() {
        let imageRect = NSRect(x: 0, y: 0, width: 16, height: 16)
        anImage = NSImage(size: imageRect.size)

        anImage?.lockFocus()
        NSGraphicsContext.current?.imageInterpolation = .high

        let path = NSBezierPath(rect: imageRect)
        fillColor.setFill()
        path.fill()

        strokeColor.setStroke()
        path.stroke()

        anImage?.unlockFocus()
    }

    @objc func colorImage() -> NSImage? {
        if anImage == nil {
            resetColorImage()
        }
        return anImage
    }

    @objc func generateGraphics() {
        // To be implemented by subclasses
    }

    @objc func draw(_: NSRect) {
        // To be implemented by subclasses
    }

    // MARK: - Notification Handlers

    @objc func geometryDidChange(_: Notification) {
        // To be implemented by subclasses
    }

    @objc func geometryDidChangePercent(_: Notification) {
        // To be implemented by subclasses
    }

    @objc func geometryDidChangeSectors(_: Notification) {
        // To be implemented by subclasses
    }

    // MARK: - Properties

    @objc func getMaxCount() -> Int {
        maxCount
    }

    @objc func getMaxPercent() -> Float {
        maxPercent
    }

    @objc func setLineWeight(_ weight: Float) {
        lineWeight = weight
    }

    @objc func getLineWeight() -> Float {
        lineWeight
    }

    @objc func setDataSet(_ set: XRDataSet?) {
        dataSet = set
    }

    @objc func getDataSet() -> XRDataSet? {
        dataSet
    }

    // MARK: - Layer Settings

    @objc func layerSettings() -> [String: Any] {
        var settings: [String: Any] = [:]
        settings["TYPE"] = Self.classTag()
        settings["VISIBLE"] = isVisible ? "YES" : "NO"
        settings["ACTIVE"] = isActive ? "YES" : "NO"
        settings["BIDIR"] = isBiDir ? "YES" : "NO"
        settings["Layer_Name"] = layerName
        settings["Line_Weight"] = String(format: "%.1f", lineWeight)
        settings["Max_Count"] = String(format: "%d", maxCount)
        settings["Max_Percent"] = String(format: "%.1f", maxPercent)
        return settings
    }

    // MARK: - XML Support

    @objc func baseXMLTree(forVersion version: String) -> LITMXMLTree {
        switch version {
        case "1.0":
            baseXMLTree1_0()
        default:
            LITMXMLTree()
        }
    }

    @objc func baseXMLTree1_0() -> LITMXMLTree {
        let tree = LITMXMLTree()
        let settings = layerSettings()
        let attributeOrder = ["TYPE", "VISIBLE", "ACTIVE", "BIDIR"]

        for key in attributeOrder {
            if let value = settings[key] as? String {
                tree.addAttribute(key, value: value)
            }
        }

        return tree
    }

    @objc class func layer(withGeometryController controller: XRGeometryController,
                           xmlTree configureTree: LITMXMLTree,
                           forVersion version: String,
                           withParentView parentView: NSView) -> XRLayer?
    {
        guard let attributes = configureTree.attributesDictionary() as? [String: String],
              let type = attributes["TYPE"]
        else {
            return nil
        }

        let layer: XRLayer?

        switch type {
        case XRLayerCore.classTag():
            layer = XRLayerCore(geometryController: controller)
        case XRLayerGrid.classTag():
            layer = XRLayerGrid(geometryController: controller)
        case XRLayerData.classTag():
            layer = XRLayerData(geometryController: controller)
        case XRLayerText.classTag():
            layer = XRLayerText(geometryController: controller)
        case XRLayerLineArrow.classTag():
            layer = XRLayerLineArrow(geometryController: controller)
        default:
            return nil
        }

        layer?.configureBase(withXMLTree: configureTree, version: version)
        layer?.configure(withXMLTree: configureTree, version: version)

        return layer
    }

    @objc func configureBase(withXMLTree configureTree: LITMXMLTree, version: String) {
        switch version {
        case "1.0":
            configureBase(withXMLTree1_0: configureTree)
        default:
            break
        }
    }

    @objc func configureBase(withXMLTree1_0 configureTree: LITMXMLTree) {
        if let attributes = configureTree.attributesDictionary() as? [String: String] {
            isVisible = attributes["VISIBLE"] == "YES"
            isActive = attributes["ACTIVE"] == "YES"
            isBiDir = attributes["BIDIR"] == "YES"
        }
    }

    @objc func configure(withXMLTree configureTree: LITMXMLTree, version: String) {
        // To be implemented by subclasses
    }

    @objc func configure(withXMLTree1_0 configureTree: LITMXMLTree) {
        // To be implemented by subclasses
    }

    @objc func xmlTree(forVersion version: String) -> LITMXMLTree {
        // To be implemented by subclasses
        LITMXMLTree()
    }
}
